FIDO Metadata verification ignores trustRootCertificate?

Using v2.4.1.

Earlier in the morning GST, FIDO metadata verification processes started to fail:

```
Caused by: java.security.cert.CertPathValidatorException: validity check failed
	at java.base/sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:135)
	at java.base/sun.security.provider.certpath.PKIXCertPathValidator.validate(PKIXCertPathValidator.java:224)
	at java.base/sun.security.provider.certpath.PKIXCertPathValidator.validate(PKIXCertPathValidator.java:144)
	at java.base/sun.security.provider.certpath.PKIXCertPathValidator.engineValidate(PKIXCertPathValidator.java:83)
	at java.base/java.security.cert.CertPathValidator.validate(CertPathValidator.java:309)
	at com.yubico.fido.metadata.FidoMetadataDownloader.verifyBlob(FidoMetadataDownloader.java:1097)
	at com.yubico.fido.metadata.FidoMetadataDownloader.parseAndVerifyBlob(FidoMetadataDownloader.java:1018)
	at com.yubico.fido.metadata.FidoMetadataDownloader.refreshBlobInternal(FidoMetadataDownloader.java:807)
	at com.yubico.fido.metadata.FidoMetadataDownloader.refreshBlob(FidoMetadataDownloader.java:795)
	... 144 more
Caused by: java.security.cert.CertificateExpiredException: NotAfter: Wed Jun 14 01:26:44 GST 2023
	at java.base/sun.security.x509.CertificateValidity.valid(CertificateValidity.java:277)
	at java.base/sun.security.x509.X509CertImpl.checkValidity(X509CertImpl.java:627)
	at java.base/sun.security.provider.certpath.BasicChecker.verifyValidity(BasicChecker.java:190)
	at java.base/sun.security.provider.certpath.BasicChecker.check(BasicChecker.java:144)
	at java.base/sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:125)
	... 154 more
```

I tracked this down to the following block in `FidoMetadataDownloader`:

```java
    if (header.getX5u().isPresent()) {
      // Download cert chain
    } else if (header.getX5c().isPresent()) {
      certChain = header.getX5c().get();
    } else {
      certChain = Collections.singletonList(trustRootCertificate);
    }
```

If I am reading this correctly, this indicates that `trustRootCertificate` potentially can never be explicitly downloaded and used for verification processes if one of the above two conditions holds true? I can probably understand that the intent here is to allow for a seamless and quick fetching of the cert chain, allowing the blob to control the cert chain but I am not sure how we can get around a failure such as the one above, if there are delays or other issues from upstream. 

If an explicit `trustRootCertificate` is configured, should it not always be used, regardless of what else exists? Or rather, use the given trust root certificate first, and only fallback if validation fails?

--- a/NEWS
+++ b/NEWS
@@ -1,11 +1,22 @@
 == Version 2.5.0 (unreleased) ==
 
+`webauthn-server-core`:
+
 New features:
 
 * Added method `.isUserVerified()` to `RegistrationResult` and `AssertionResult`
   as a shortcut for accessing the UV flag in authenticator data.
 * Updated README and JavaDoc to use the "passkey" term and provide more guidance
   around passkey use cases.
+
+`webauthn-server-attestation`:
+
+New features:
+
+* Added option `verifyDownloadsOnly(boolean)` to `FidoMetadataDownloader`. When
+  set to `true`, the BLOB signature will not be verified when loading a BLOB
+  from cache or when explicitly given. Default setting is `false`, which
+  preserves the previous behaviour.
 
 
 == Version 2.4.1 ==

--- a/webauthn-server-attestation/README.adoc
+++ b/webauthn-server-attestation/README.adoc
@@ -154,6 +154,8 @@ FidoMetadataDownloader downloader = FidoMetadataDownloader.builder()
   .useTrustRootCacheFile(new File("/var/cache/webauthn-server/fido-mds-trust-root.bin"))
   .useDefaultBlob()
   .useBlobCacheFile(new File("/var/cache/webauthn-server/fido-mds-blob.bin"))
+  .verifyDownloadsOnly(true)  // Recommended, otherwise cache may expire if BLOB certificate expires
+                              // See: https://github.com/Yubico/java-webauthn-server/issues/294
   .build();
 
 FidoMetadataService mds = FidoMetadataService.builder()

--- a/webauthn-server-attestation/src/main/java/com/yubico/fido/metadata/FidoMetadataDownloader.java
+++ b/webauthn-server-attestation/src/main/java/com/yubico/fido/metadata/FidoMetadataDownloader.java
@@ -87,6 +87,7 @@ import lombok.AccessLevel;
 import lombok.AllArgsConstructor;
 import lombok.NonNull;
 import lombok.RequiredArgsConstructor;
+import lombok.Value;
 import lombok.extern.slf4j.Slf4j;
 
 /**
@@ -119,6 +120,7 @@ public final class FidoMetadataDownloader {
   private final CertStore certStore;
   @NonNull private final Clock clock;
   private final KeyStore httpsTrustStore;
+  private final boolean verifyDownloadsOnly;
 
   /**
    * Begin configuring a {@link FidoMetadataDownloader} instance. See the {@link
@@ -148,6 +150,7 @@ public final class FidoMetadataDownloader {
     private CertStore certStore = null;
     @NonNull private Clock clock = Clock.systemUTC();
     private KeyStore httpsTrustStore = null;
+    private boolean verifyDownloadsOnly = false;
 
     public FidoMetadataDownloader build() {
       return new FidoMetadataDownloader(
@@ -165,7 +168,8 @@ public final class FidoMetadataDownloader {
           blobCacheConsumer,
           certStore,
           clock,
-          httpsTrustStore);
+          httpsTrustStore,
+          verifyDownloadsOnly);
     }
 
     /**
@@ -611,6 +615,26 @@ public final class FidoMetadataDownloader {
 
       return this;
     }
+
+    /**
+     * If set to <code>true</code>, the BLOB signature will not be verified when loading the BLOB
+     * from cache or when explicitly set via {@link Step4#useBlob(String)}. This means that if a
+     * BLOB was successfully verified once and written to cache, that cached value will be
+     * implicitly trusted when loaded in the future.
+     *
+     * <p>If set to <code>false</code>, the BLOB signature will always be verified no matter where
+     * the BLOB came from. This means that a cached BLOB may become invalid if the BLOB certificate
+     * expires, even if the BLOB was successfully verified at the time it was downloaded.
+     *
+     * <p>The default setting is <code>false</code>.
+     *
+     * @param verifyDownloadsOnly <code>true</code> if the BLOB signature should be ignored when
+     *     loading the BLOB from cache or when explicitly set via {@link Step4#useBlob(String)}.
+     */
+    public FidoMetadataDownloaderBuilder verifyDownloadsOnly(final boolean verifyDownloadsOnly) {
+      this.verifyDownloadsOnly = verifyDownloadsOnly;
+      return this;
+    }
   }
 
   /**
@@ -960,7 +984,7 @@ public final class FidoMetadataDownloader {
           FidoMetadataDownloaderException {
     if (blobJwt != null) {
       return Optional.of(
-          parseAndVerifyBlob(
+          parseAndMaybeVerifyBlob(
               new ByteArray(blobJwt.getBytes(StandardCharsets.UTF_8)), trustRootCertificate));
 
     } else {
@@ -987,7 +1011,7 @@ public final class FidoMetadataDownloader {
     return cachedContents.map(
         cached -> {
           try {
-            return parseAndVerifyBlob(cached, trustRootCertificate);
+            return parseAndMaybeVerifyBlob(cached, trustRootCertificate);
           } catch (Exception e) {
             log.warn("Failed to read or parse cached BLOB.", e);
             return null;
@@ -1044,18 +1068,27 @@ public final class FidoMetadataDownloader {
           InvalidKeyException,
           Base64UrlException,
           FidoMetadataDownloaderException {
-    Scanner s = new Scanner(new ByteArrayInputStream(jwt.getBytes())).useDelimiter("\\.");
-    final ByteArray header = ByteArray.fromBase64Url(s.next());
-    final ByteArray payload = ByteArray.fromBase64Url(s.next());
-    final ByteArray signature = ByteArray.fromBase64Url(s.next());
-    return verifyBlob(header, payload, signature, trustRootCertificate);
+    return verifyBlob(parseBlob(jwt), trustRootCertificate);
   }
 
-  private MetadataBLOB verifyBlob(
-      ByteArray jwtHeader,
-      ByteArray jwtPayload,
-      ByteArray jwtSignature,
-      X509Certificate trustRootCertificate)
+  private MetadataBLOB parseAndMaybeVerifyBlob(ByteArray jwt, X509Certificate trustRootCertificate)
+      throws CertPathValidatorException,
+          InvalidAlgorithmParameterException,
+          CertificateException,
+          IOException,
+          NoSuchAlgorithmException,
+          SignatureException,
+          InvalidKeyException,
+          Base64UrlException,
+          FidoMetadataDownloaderException {
+    if (verifyDownloadsOnly) {
+      return parseBlob(jwt).blob;
+    } else {
+      return verifyBlob(parseBlob(jwt), trustRootCertificate);
+    }
+  }
+
+  private MetadataBLOB verifyBlob(ParseResult parseResult, X509Certificate trustRootCertificate)
       throws IOException,
           CertificateException,
           NoSuchAlgorithmException,
@@ -1064,12 +1097,7 @@ public final class FidoMetadataDownloader {
           CertPathValidatorException,
           InvalidAlgorithmParameterException,
           FidoMetadataDownloaderException {
-    final ObjectMapper headerJsonMapper =
-        com.yubico.internal.util.JacksonCodecs.json()
-            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, true)
-            .setBase64Variant(Base64Variants.MIME_NO_LINEFEEDS);
-    final MetadataBLOBHeader header =
-        headerJsonMapper.readValue(jwtHeader.getBytes(), MetadataBLOBHeader.class);
+    final MetadataBLOBHeader header = parseResult.blob.getHeader();
 
     final List<X509Certificate> certChain;
     if (header.getX5u().isPresent()) {
@@ -1117,9 +1145,9 @@ public final class FidoMetadataDownloader {
 
     signature.initVerify(leafCert.getPublicKey());
     signature.update(
-        (jwtHeader.getBase64Url() + "." + jwtPayload.getBase64Url())
+        (parseResult.jwtHeader.getBase64Url() + "." + parseResult.jwtPayload.getBase64Url())
             .getBytes(StandardCharsets.UTF_8));
-    if (!signature.verify(jwtSignature.getBytes())) {
+    if (!signature.verify(parseResult.jwtSignature.getBytes())) {
       throw new FidoMetadataDownloaderException(Reason.BAD_SIGNATURE);
     }
 
@@ -1134,10 +1162,28 @@ public final class FidoMetadataDownloader {
     pathParams.setDate(Date.from(clock.instant()));
     cpv.validate(blobCertPath, pathParams);
 
-    return new MetadataBLOB(
-        header,
-        JacksonCodecs.jsonWithDefaultEnums()
-            .readValue(jwtPayload.getBytes(), MetadataBLOBPayload.class));
+    return parseResult.blob;
+  }
+
+  private static ParseResult parseBlob(ByteArray jwt) throws IOException, Base64UrlException {
+    Scanner s = new Scanner(new ByteArrayInputStream(jwt.getBytes())).useDelimiter("\\.");
+    final ByteArray jwtHeader = ByteArray.fromBase64Url(s.next());
+    final ByteArray jwtPayload = ByteArray.fromBase64Url(s.next());
+    final ByteArray jwtSignature = ByteArray.fromBase64Url(s.next());
+
+    final ObjectMapper headerJsonMapper =
+        com.yubico.internal.util.JacksonCodecs.json()
+            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, true)
+            .setBase64Variant(Base64Variants.MIME_NO_LINEFEEDS);
+
+    return new ParseResult(
+        new MetadataBLOB(
+            headerJsonMapper.readValue(jwtHeader.getBytes(), MetadataBLOBHeader.class),
+            JacksonCodecs.jsonWithDefaultEnums()
+                .readValue(jwtPayload.getBytes(), MetadataBLOBPayload.class)),
+        jwtHeader,
+        jwtPayload,
+        jwtSignature);
   }
 
   private static ByteArray readAll(InputStream is) throws IOException {
@@ -1157,5 +1203,13 @@ public final class FidoMetadataDownloader {
     } else {
       return null;
     }
+  }
+
+  @Value
+  private static class ParseResult {
+    private MetadataBLOB blob;
+    private ByteArray jwtHeader;
+    private ByteArray jwtPayload;
+    private ByteArray jwtSignature;
   }
 }

--- a/webauthn-server-attestation/src/test/scala/com/yubico/fido/metadata/FidoMetadataDownloaderSpec.scala
+++ b/webauthn-server-attestation/src/test/scala/com/yubico/fido/metadata/FidoMetadataDownloaderSpec.scala
@@ -39,11 +39,13 @@ import java.security.SecureRandom
 import java.security.cert.CRL
 import java.security.cert.CertPathValidatorException
 import java.security.cert.CertPathValidatorException.BasicReason
+import java.security.cert.CertificateExpiredException
 import java.security.cert.X509Certificate
 import java.time.Clock
 import java.time.Instant
 import java.time.LocalDate
 import java.time.ZoneOffset
+import java.time.temporal.ChronoUnit
 import java.util.Optional
 import javax.servlet.http.HttpServletRequest
 import javax.servlet.http.HttpServletResponse
@@ -157,16 +159,19 @@ class FidoMetadataDownloaderSpec
     certs
   }
 
+  private def formatJwtTbs(header: String, body: String): String =
+    new ByteArray(
+      header.getBytes(StandardCharsets.UTF_8)
+    ).getBase64Url + "." + new ByteArray(
+      body.getBytes(StandardCharsets.UTF_8)
+    ).getBase64Url
+
   private def makeBlob(
       blobKeypair: KeyPair,
       header: String,
       body: String,
   ): String = {
-    val blobTbs = new ByteArray(
-      header.getBytes(StandardCharsets.UTF_8)
-    ).getBase64Url + "." + new ByteArray(
-      body.getBytes(StandardCharsets.UTF_8)
-    ).getBase64Url
+    val blobTbs = formatJwtTbs(header, body)
     val blobSignature = TestAuthenticator.sign(
       new ByteArray(blobTbs.getBytes(StandardCharsets.UTF_8)),
       blobKeypair.getPrivate,
@@ -193,6 +198,22 @@ class FidoMetadataDownloaderSpec
       "entries": []
     }"""
     makeBlob(blobKeypair, blobHeader, blobBody)
+  }
+
+  private def makeUnsignedBlob(
+      nextUpdate: LocalDate,
+      legalHeader: String = "Kom ihåg att du aldrig får snyta dig i mattan!",
+      no: Int = 1,
+  ): String = {
+    val blobHeader =
+      s"""{"alg":"None"}"""
+    val blobBody = s"""{
+      "legalHeader": "${legalHeader}",
+      "no": ${no},
+      "nextUpdate": "${nextUpdate}",
+      "entries": []
+    }"""
+    formatJwtTbs(blobHeader, blobBody) + ".AAAA"
   }
 
   private def makeHttpServer(
@@ -1847,6 +1868,158 @@ class FidoMetadataDownloaderSpec
           ).getPayload
           blob should not be null
           blob.getNo should equal(oldBlobNo)
+        }
+
+        it("If verifyDownloadsOnly is not set, a cached BLOB may expire.") {
+          val oldBlobNo = 1
+          val newBlobNo = 2
+
+          val (trustRootCert, caKeypair, caName) = makeTrustRootCert()
+          val (blobCert, blobKeypair, _) = makeCert(caKeypair, caName)
+          val oldBlobJwt =
+            makeBlob(
+              List(blobCert),
+              blobKeypair,
+              CertValidTo.atOffset(ZoneOffset.UTC).toLocalDate,
+              no = oldBlobNo,
+            )
+          val newBlobJwt =
+            makeBlob(
+              List(blobCert),
+              blobKeypair,
+              CertValidTo.atOffset(ZoneOffset.UTC).toLocalDate,
+              no = newBlobNo,
+            )
+          val crls = List[CRL](
+            TestAuthenticator.buildCrl(
+              caName,
+              caKeypair.getPrivate,
+              "SHA256withECDSA",
+              CertValidFrom,
+              CertValidTo,
+            )
+          )
+
+          val (server, serverUrl, httpsCert) =
+            makeHttpServer("/blob.jwt", newBlobJwt)
+          startServer(server)
+
+          val thrown = the[CertPathValidatorException] thrownBy {
+            load(
+              FidoMetadataDownloader
+                .builder()
+                .expectLegalHeader(
+                  "Kom ihåg att du aldrig får snyta dig i mattan!"
+                )
+                .useTrustRoot(trustRootCert)
+                .downloadBlob(new URL(s"${serverUrl}/blob.jwt"))
+                .useBlobCache(
+                  () =>
+                    Optional.of(
+                      new ByteArray(oldBlobJwt.getBytes(StandardCharsets.UTF_8))
+                    ),
+                  _ => {},
+                )
+                .clock(
+                  Clock
+                    .fixed(CertValidTo.plus(1, ChronoUnit.DAYS), ZoneOffset.UTC)
+                )
+                .useCrls(crls.asJava)
+                .trustHttpsCerts(httpsCert)
+                .build()
+            ).getPayload
+          }
+          thrown.getCause shouldBe a[CertificateExpiredException]
+        }
+
+        it("If verifyDownloadsOnly is set, the signature is ignored when loading from cache.") {
+          val oldBlobNo = 1
+          val newBlobNo = 2
+
+          val (trustRootCert, caKeypair, caName) = makeTrustRootCert()
+          val (blobCert, blobKeypair, _) = makeCert(caKeypair, caName)
+          val oldBlobJwt =
+            makeUnsignedBlob(
+              CertValidFrom.atOffset(ZoneOffset.UTC).toLocalDate,
+              no = oldBlobNo,
+            )
+          val newBlobJwt =
+            makeBlob(
+              List(blobCert),
+              blobKeypair,
+              CertValidTo.atOffset(ZoneOffset.UTC).toLocalDate,
+              no = newBlobNo,
+            )
+          val crls = List[CRL](
+            TestAuthenticator.buildCrl(
+              caName,
+              caKeypair.getPrivate,
+              "SHA256withECDSA",
+              CertValidFrom,
+              CertValidTo,
+            )
+          )
+
+          val (server, serverUrl, httpsCert) =
+            makeHttpServer("/blob.jwt", newBlobJwt)
+          startServer(server)
+
+          val blob = load(
+            FidoMetadataDownloader
+              .builder()
+              .expectLegalHeader(
+                "Kom ihåg att du aldrig får snyta dig i mattan!"
+              )
+              .useTrustRoot(trustRootCert)
+              .downloadBlob(new URL(s"${serverUrl}/blob.jwt"))
+              .useBlobCache(
+                () =>
+                  Optional.of(
+                    new ByteArray(oldBlobJwt.getBytes(StandardCharsets.UTF_8))
+                  ),
+                _ => {},
+              )
+              .clock(
+                Clock.fixed(
+                  CertValidTo.plus(1, ChronoUnit.DAYS),
+                  ZoneOffset.UTC,
+                )
+              )
+              .useCrls(crls.asJava)
+              .trustHttpsCerts(httpsCert)
+              .verifyDownloadsOnly(true)
+              .build()
+          ).getPayload
+          blob should not be null
+          blob.getNo should equal(oldBlobNo)
+        }
+
+        it("If verifyDownloadsOnly is set, the signature is ignored when loading an explicitly given BLOB.") {
+          val blobNo = 1
+          val (trustRootCert, caKeypair, caName) = makeTrustRootCert()
+          val blobJwt =
+            makeUnsignedBlob(
+              CertValidFrom.atOffset(ZoneOffset.UTC).toLocalDate,
+              no = blobNo,
+            )
+
+          val blob = load(
+            FidoMetadataDownloader
+              .builder()
+              .expectLegalHeader(
+                "Kom ihåg att du aldrig får snyta dig i mattan!"
+              )
+              .useTrustRoot(trustRootCert)
+              .useBlob(blobJwt)
+              .clock(
+                Clock
+                  .fixed(CertValidTo.plus(1, ChronoUnit.DAYS), ZoneOffset.UTC)
+              )
+              .verifyDownloadsOnly(true)
+              .build()
+          ).getPayload
+          blob should not be null
+          blob.getNo should equal(blobNo)
         }
       }
 

